### PR TITLE
add instrument for handler.sendMessage(); fix getAllHandlers() bug

### PR DESCRIPTION
**Pull request by Canonize**

### CODE REVIEW

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/MessageHandler.java [hunk 1]
```diff
@@ -28,7 +28,7 @@ public Set<SootClass> getAllHandlers() {
 			for (Iterator<SootClass> iter = Scene.v().getApplicationClasses().snapshotIterator(); iter.hasNext();) {
 				SootClass sootClass = iter.next();
 
-				SootClass tmpClass = sootClass.getSuperclassUnsafe();
+				SootClass tmpClass =  Scene.v().getSootClass(sootClass.getName());
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/MessageHandler.java]: This is equal to ``tmpClass = sootClass``. The original code obtained the superclass. Please check,

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java [hunk 2]
```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
...
+						//+++ collect handler fields signature	
+						if (stmt.containsFieldRef() && stmt.getFieldRef().getType().toString().contains("android.os.Handler")) {
+							if(stmt.getUseBoxes().size() > 1 && handlerInner.get(stmt.getFieldRef().getField().getSignature()) == null)
+								handlerInner.put(stmt.getFieldRef().getField().getSignature(),stmt.getUseBoxes().get(1).getValue().getType().toString());
+							// use value as locator
+							handlerClass.put(sootClass.getName() + ":" + stmt.getDefBoxes().get(0).getValue(),stmt.getFieldRef().getField().getSignature());
+						}
+							
+						if(stmt.containsInvokeExpr() && stmt.getInvokeExpr().getMethodRef().getSignature().contains("android.os.Handler")) {
+							//Soot has a access$XXX method that returns private fields in outer class
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: The ``access`` methods are part of the bytecode. They are generated by the Java compiler to allow access to private methods from inner classes (and some other special cases). They are not an artifact of Soot. The code must work regardless of whether there is an ``access`` method or a direct call. Do not rely on the structure of the ``access`` method.

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java [hunk 3]
```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
...
-									SootMethod initMethod = handler.getMethod("void <init>()");
-									Unit initU = Jimple.v().newInvokeStmt(
-											Jimple.v().newSpecialInvokeExpr(handlerLocal, initMethod.makeRef()));
-									initU.addTag(SimulatedCodeElementTag.TAG);
-									body.getUnits().insertAfter(initU, newU);
-									instrumentedUnits.put(body, initU);
+						Unit unit = unitIter.next();
+						Stmt stmt = (Stmt) unit;
+						//+++ collect handler fields signature	
+						if (stmt.containsFieldRef() && stmt.getFieldRef().getType().toString().contains("android.os.Handler")) {
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: Please check whether the result of ``FieldRef.getType`` is a ``RefType``. On a ``RefType``, you can call ``getSootClass().getName()`` to get the proper name of the class. Don't expect ``toString()`` to be anything but debug information.

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java [hunk 4]
```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
...
+							if(stmt.getUseBoxes().size() > 1 && handlerInner.get(stmt.getFieldRef().getField().getSignature()) == null)
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: ``Map.putIfAbsent())`` is more efficient and easier to read.

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java [hunk 5]
```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
...
+							//Soot has a access$XXX method that returns private fields in outer class
+							if(stmt.getInvokeExpr().getMethodRef().getName().contains("access")) {
+								Body b = stmt.getInvokeExpr().getMethod().retrieveActiveBody();
+								for (Iterator<Unit> u = b.getUnits().iterator(); u.hasNext();) {
+									Stmt s = (Stmt) u.next();
+									//if the access method return a handler, collect its signature
+									if(s.containsFieldRef() && s.getFieldRef().getType().toString().contains("android.os.Handler"))
+									{
+										if(!stmt.getDefBoxes().isEmpty())
+											handlerClass.put(sootClass.getName() + ":" + stmt.getDefBoxes().get(0).getValue(),s.getFieldRef().getField().getSignature());
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: Do not use string splitting. Use a proper object to put into the map.

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java [hunk 6]
```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
...
+									}
+								}
+							}
+						}													
+					}
+
+					body.validate();
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: This should not be done by default, because it can take a lot of time.

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java [hunk 7]
```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
...
+				final LocalGenerator lg = Scene.v().createLocalGenerator(body);
+				
+				for (Iterator<Unit> unitIter = body.getUnits().snapshotIterator(); unitIter.hasNext();) {
+					Unit unit = unitIter.next();
+					Stmt stmt = (Stmt) unit;
+					if (stmt.containsInvokeExpr()) {
+						SootMethod callee = stmt.getInvokeExpr().getMethod();
+						
+						// For sendMessage(), we directly call the respective handler.handleMessage()
+						if (callee == smMessengerSend || callee == Scene.v().grabMethod("<android.os.Handler: boolean sendMessage(android.os.Message)>")) {
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: This method only needs to be resolved once, instead of resolving it again in every iteration of the loop.

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java [hunk 8]
```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
...
+							//collect the value for sendMessage()
+							String hc = handlerClass.get(sootClass.getName() + ":" + stmt.getInvokeExpr().getUseBoxes().get(1).getValue().toString());
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: As stated above, you should have a proper object that stores the Soot objects instead of reyling on string parsing.

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java [hunk 9]
```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
...
+							//System.out.println("--1: "+hc);
+							Set<SootClass> handlers = MessageHandler.v().getAllHandlers();								
+							soot.javaToJimple.InitialResolver.v().getNextPrivateAccessCounter();
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: Why is this line needed?

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java [hunk 10]
```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
...
 						}
-					}
-
-					body.validate();
+					}	
+												
+					
 				}
 
+				body.validate();
```
Comment (StevenArzt) [inline soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: As above, this is nice for debugging, but too slow for productive code.

**Review by StevenArzt — CHANGES_REQUESTED** [review]

File: soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java (hunk 2 above)
Comment (Canonize) [reply soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java]: Thanks for your patient reply, sorry for not getting back to you sooner. 
This code is meant to deal with private handlers. Removing it would be fine. 
BTW, if not rely on the structure of the `access` method, how does soot deal with Private handlers/inner classes?
